Rework input widgets

theGreatWhiteShark opened the pull request:
Update:

This PR has four main goals:

### 1. Improve scalability

- most PNG files have been either replaced by a SVG one or are drawn directly using the Qt paint engine
- labels "hardcoded" in the PNG files have been replaced by subclassed `QLabel`s
- rendering of spin boxes, buttons, and combo boxes is now directly done by Qt itself since I threw away most of the custom widget code and based them on `QDoubleSpinBox`, `QComboBox`, and `QPushButton` instead.

### 2. Improve internationalization

- every single label in the GUI should be translatable now
- every translatable object does support UTF8 characters now (including the status display in the PlayerControl)

### 3. Improve accessibility

- increased widget size to fill available space
- PreferencesDialog option to increase the font size for all labels and button text (I'm even able to use it without glasses now :D )
- introduced possibility to increase the color contrast - almost all colors are exposed in the PreferencesDialog

### 4. Improve usability

- by basing the widgets on Qt's default ones their UX should feel way more similar to the one found in other programs
- all input widgets do support mouse and keyboard inputs now
- since only one widget - the one having focus - can accept keyboard input, all widgets capable of doing so are faintly highlighted when hovered and strongly highlighted when focused
- if a widget is midi-learnable, this fact is indicated by its tooltip

Apart from that **this PR is not an attempt to improve the design and styling**. I tried to make things resemble the previous look as closely as possible. But I had to drop the former light blue color e.g. the spin boxes because they just didn't look nice. But I'm open to change the default values of the individual colors. On the other hand, this PR makes changing the styling of Hydrogen way more easy (editing CSS-like stylesheets instead of tweaking pixels).


----

Former text:

Rework of the following classes

- [x] `Rotary`
- [x] `Knob`
- [x] `*Fader`
- [x] `LCD`

### What has been changed

* images are now based on SVG (with room for artistic improvement)
* widgets can be de- and activated from within the code while visibly representing this state
* mouse move and scrolling to update the values (also for the LCDs)
* after clicking on the widgets the following keyboard input is supported
  * <kbd>up / right</kbd> - increase value 
  * <kbd>down / left</kbd> - decrease value 
  * <kbd>Ctrl + up / right</kbd> - faster increase value 
  * <kbd>Ctrl + down / left</kbd> - faster decrease value 
  * <kbd>page up</kbd> - faster increase value 
  * <kbd>page down</kbd> - faster decrease value 
  * <kbd>home</kbd> - set maximum value
  * <kbd>end</kbd> - set minimum value
  * Using the number keys + <kbd>.</kbd> and <kbd>-</kbd> the user can input the desired value directly. Updates will be performed immediately. E.g. entering <kdb>0</kdb> will set the value to `0`, followed by <kbd>.</kbd> - still `0`, followed by <kbd>2</kbd> - `0.2` set ... After two seconds without user input the internal input buffer will reset.
* as soon as a widget is hovered by the mouse it will be underlined (currently blue (activated) or grey (deactivated)) 
* the current values of all widgets will be shown in the tooltip alongside the tooltip text as well as in a tooltip-like text while altering their values (this will not apply for the LCDs since they already show their values.) As a consequence the `RotaryTooltip` class died in order to make things consistent.

Did I miss something that should be implemented too? 

### Notes

* Clicking a widget to allow keyboard input is required right now because I didn't want to mess with our focus/tab policy (yet)
* After finishing this PR I will squash its commits since there are iterative changes to the pixmap graphics which would add unnecessary size to the repo.

addresses #863

oddtime commented:
I like the new features very much!

The fine pitch widgets are shown green if the value is in [-0.1;0.1] etc.
Is it wanted?

theGreatWhiteShark commented:
> The fine pitch widgets are shown green if the value is in [-0.1;0.1] etc.
Is it wanted?

Thanks a lot. There were actually a number of things I wanted discuss with you people in here (man, I start to make notes as the times where I could have kept that in mind seems to are gone).

1. One of these is indeed the green center of the centered knobs. Previously there was a limited set of pixmaps and a range of values was covered by the one showing the green mid position. Now, only the rendering engine and overall scaling factor of the application impose a constraint in the possible angles and I used green for the mean value +/- 1 percent of the full range to mimik the previous behavior. One of the reasons for this PR is that I use a Asus laptop (with spontaneous touchpad glitches on hardware level) without a mouse. So, setting specific values requires a lot of patience. I thought about making the center sticky to emulate the additional resistance found on instruments or amp knobs. But upon rethinking this idea I dismissed it as overly complex and unintuitive. We could also require the default value of centered knobs to be the center so one can use <kbd>shift + left click</kbd> to reset the widget to the center and make just this very value be displayed green. But I'm not sure how many users are familiar with this feature.
2. The current value of a rotary is now indicated using a circle instead of a line. The code for the latter is pushed as well and commented out in the `Rotary::paintEvent()`. But I found it too render quit badly with the small sizes we use for the widgets.
3. A widget will now be underlined when hovering over it. I introduced this to explicitly mark the widget as interactive and to indicate (using different colors) whether it is handling user input or not. But it gives the wrong idea about when the widget will accept keyboard input because it has to be clicked first. Well, while writing this down I think I will introduce a third state of underlying: none, slight and transparent color for hovering, and strong color for clicked.
4. To be consistent with the previous design, I added the rendering of the limits in the bottom right and left corners of the rotary too. It's either an integer number in between (-10, 10) or a + or - otherwise. I tried to render numbers >= 10 but it doesn't look good. I'm not quite sure whether these are of any use at all and maybe we can drop them at some point (because they are not visually appealing). Instead, we could e.g. not the upper and lower bound in the tooltip of the widget.

oddtime commented:
I think that rotary should display green light only for the precise center value.

> I thought about making the center sticky to emulate the additional resistance found on instruments or amp knobs. But upon rethinking this idea I dismissed it as overly complex and unintuitive.

Yes, personally I would prefer it linear.

> The current value of a rotary is now indicated using a circle instead of a line. The code for the latter is pushed as well and commented out in the Rotary::paintEvent(). But I found it too render quit badly with the small sizes we use for the widgets.

That's a thing I noticed; probably I preferred the lines but maybe because I was used to see them.
I find also on my screen that the circle is not rendered perfectly aligned with the green central led.

> We could also require the default value of centered knobs to be the center so one can use shift + left click to reset the widget to the center and make just this very value be displayed green. But I'm not sure how many users are familiar with this feature.

I have been desiring a way to reset the central rotary widgets for long time and discovered I could do it with ctrl+click just few weeks ago.

One unexpected thing for me was that I had to wait a time to reenter the value with the number keys (if you enter a too big number and then all the further entries add to the buffers but the user doesn't know, if I understand it).
Have you already thought to make a temporary display to appear to show the buffer and exit with Enter? Initially I thought that this was good after double click. But if that takes a long time to implement the current way is good to me.

theGreatWhiteShark commented:
> That's a thing I noticed; probably I preferred the lines but maybe because I was used to see them.

You can have a look at the version with the lines by commenting out 

```C++
	if ( m_bIsActive ) {
		float fDistance, fRadius, fBaseX, fBaseY;
		if ( m_type == Type::Small ) {
			fBaseX = 9.0;
			fBaseY = 9.0;
			fDistance = 3;
			fRadius = 1;
		} else {
			fBaseX = 22.0;
			fBaseY = 14.0;
			fDistance = 4;
			fRadius = 1.5;
		}

		QPointF p1( fBaseX + std::cos( fCurrentAngle ) * fDistance,
					fBaseY + std::sin( fCurrentAngle ) * fDistance );
		painter.setPen( QPen( colorHandle, 1 ) );
		painter.setBrush( QBrush( colorHandle ) );
		painter.drawEllipse( p1, fRadius, fRadius );
	}
```

in the `paintEvent()` function in `src/gui/src/Widgets/Rotary.cpp` and uncomment 

```c++
	
	float fLength, fWidth, fBaseX, fBaseY;
	if ( m_type == Type::Small ) {
		fBaseX = 9.0;
		fBaseY = 9.0;
		fLength = 4;
		fWidth = 2;
	} else {
		fBaseX = 22.0;
		fBaseY = 14.0;
		fLength = 6;
		fWidth = 3;
	}
	
	QPointF p1( fBaseX + std::cos( fCurrentAngle + fPi / 2 ) * fWidth / 2,
				fBaseY + std::sin( fCurrentAngle + fPi / 2 ) * fWidth / 2 );
	QPointF p2( p1.x() + std::cos( fCurrentAngle ) * fLength,
				p1.y() + std::sin( fCurrentAngle ) * fLength );
	QPointF p3( p2.x() - std::cos( fCurrentAngle + fPi / 2 ) * fWidth / 2,
				p2.y() - std::sin( fCurrentAngle + fPi / 2 ) * fWidth / 2 );
	QPointF p4( p3.x() - std::cos( fCurrentAngle ) * fLength,
				p3.y() - std::sin( fCurrentAngle ) * fLength );
	QPainterPath path;
	path.moveTo( p1 );
	path.lineTo( p2 );
	path.lineTo( p3 );
	path.lineTo( p4 );
	
	path.setFillRule( Qt::WindingFill );
	QPen pen( colorHandle );
	pen.setJoinStyle( Qt::RoundJoin );
	pen.setWidth( 1.7 );
	painter.setPen( pen );
	painter.setBrush( QBrush( colorHandle ) );
	painter.drawPath( path );
```

> I find also on my screen that the circle is not rendered perfectly aligned with the green central led.

Okay. I'll have a look. (Strangely the widths and heights I measures in Inkscape do not work with the rendering in Qt. That's why I ended up tuning a lot of stuff by hand)

> One unexpected thing for me was that I had to wait a time to reenter the value with the number keys (if you enter a too big number and then all the further entries add to the buffers but the user doesn't know, if I understand it).

Exactly. Right now one has to wait for 2s till the input resets. I choose it as a tradeoff between the need to correct typos fast and allowing the users to enter the values slowly and calmly. Maybe a press of <kbd>ESC</kdb> to reset the input can be added. Would you expect such a key/feature to be present?

Yes, right now one can input the value with float precision. I thought about allowing only values up to the second digit after decimal point but didn't see a point in this restriction.

> Have you already thought to make a temporary display to appear to show the buffer and exit with Enter?

That's a very good idea! Using a static tooltip to show the input (also when in- and decreasing the value using the keyboard in general) will help a lot. Let's see if I can add this. Just tried it for 1-2 hours and the x/y position of the widget doesn't place the text where I want it to be.

oddtime commented on src/gui/src/Widgets/Rotary.cpp:
```diff
@@ -20,324 +20,247 @@
-{
-	if ( fDefaultValue == m_fDefaultValue ) {
-		return;
+	float fPi = std::acos( -1 );
```
Why aren't you using `M_PI` from `math.h`?

theGreatWhiteShark replied:
I've read on Stackoverflow that `M_PI` is not defined in the C++ standard itself and platform specific.

oddtime commented:
> Strangely the widths and heights I measures in Inkscape do not work with the rendering in Qt

Maybe because pixel position are considered ordinal numbers? I don't know.

oddtime commented:
I am playing a little bit with the rendering... how did you measure the min and max rotary angles? it seems to me near [225 deg, -45 deg] or slightly less extended. I am referring to angle as in a trigonometric system
			y	
			|  /
			| /  
			|/) angle
			----------> x

oddtime commented:
At start, the tooltip for fine pitch tells 0.5 rather than 0.0

oddtime commented:
> Maybe a press of ESC to reset the input can be added. Would you expect such a key/feature to be present?

I wouldn't have expected but I like it and it is reasonable

theGreatWhiteShark commented:
> Maybe because pixel position are considered ordinal numbers?

Don't know. When just looking in the docs of Qt5 there shouldn't be anything out of the ordinary.

> I am playing a little bit with the rendering... how did you measure the min and max rotary angles? it seems to me near [225 deg, -45 deg] or slightly less extended. I am referring to angle as in a trigonometric system

You are right. Also, I used some code of the [Carla project](https://github.com/falkTX/Carla/blob/main/source/frontend/widgets/scalabledial.py#L415) as first draft and did some graphical finetuning afterwards. As you have written zero degree is on the horizontal axis to the right and positive angles move counter clockwise (at least for me this is confusing as hell).

> At start, the tooltip for fine pitch tells 0.5 rather than 0.0

Indeed. I'll have a look.

theGreatWhiteShark commented:
I updated the buttons as well and took the liberty to change the design of some of the icons, like the 'select mode' button  

![new_buttons](https://user-images.githubusercontent.com/14164547/120083761-56acee80-c0cb-11eb-9666-39903236dfbf.png)

(I have no idea what "CLEAR" is in Cantonese. Just wanted to show off a bit since we now support translation in all buttons (even the ones holding only a single character))

oddtime commented:
> You are right. Also, I used some code of the Carla project as first draft and did some graphical finetuning afterwards.

Ok, when I tried rendering the line I found that it was more [210 deg, -30 deg] to be aligned to the red arc.
But the rendering of the polygon was not very good for me too at certain angles, even after closing the last side of polygon from p4 to p1.

> As you have written zero degree is on the horizontal axis to the right and positive angles move counter clockwise (at least for me this is confusing as hell).

What still confuses me is the y axis which points toward the bottom unlike the way it is commonly drawn! XD


> Maybe because pixel position are considered ordinal numbers?
> 
> Don't know. When just looking in the docs of Qt5 there shouldn't be anything out of the ordinary.

Ok, since is the knob width is an even number of pixel, I thought that maybe the center ( what is called xbase) should be between the two central pixels (in the edge of the pixel). Again, this was just a thought.

oddtime commented:
What if we make some font size bigger?
Mixer, Instrument Rack, Layers...

theGreatWhiteShark commented:
> What if we make some font size bigger?
Mixer, Instrument Rack, Layers...

Sure. I'll wanted to merge #1271 first since this will make tweaking the button font much more easy.

> But the rendering of the polygon was not very good for me too at certain angles, even after closing the last side of polygon from p4 to p1.

I tried a couple of other geometries too, like a rectangle joined by a semicircle at each end. But the line section was never satisfying due to the small size of the widgets.

> Ok, since is the knob width is an even number of pixel, I thought that maybe the center ( what is called xbase) should be between the two central pixels (in the edge of the pixel). Again, this was just a thought.

Unfortunately, it's not really on a pixel basis anymore. Both the geometries within the SVG and some of the rectangles Qt's painter does draw in are of floating point position. Depending on the scaling the Qt5 drawing/rendering engine coarse grains the afterwards.

theGreatWhiteShark commented:
I'm done replacing all pixmap labels with scalable ones.

theGreatWhiteShark commented:
@mauser I've integrate the MIDI binding into the tooltip of the widgets. Format is as follows

> $DESCRIPTION: $CURRENT_VALUE
> 
> Range: [$MIN, $MAX]
> MIDI: $ACTION  bound to [$MIDI_EVENT : $PARAMETER]

or 

> $DESCRIPTION: $CURRENT_VALUE
> 
> Range: [$MIN, $MAX]
> MIDI: $ACTION  not bound

or 

> $DESCRIPTION: $CURRENT_VALUE
> 
> Range: [$MIN, $MAX]
> This element is not MIDI operable.

depending on the state of the widget. Is this fine?

theGreatWhiteShark commented:
Hey @oddtime @cme ,

I've started looking at the size widget in the PatternEditor since the `LCDDisplay` class will die in this PR.

Right now I see two straight forward ways to replace it

1. An `QLineEdit`, like the one used when clicking the current LCDDisplay
2. or two separate LCDSpinBox widgets as shown below (just a mockup. Code is pushed but the widget is not "wired" yet.)

![size_widget](https://user-images.githubusercontent.com/14164547/121953773-45681100-cd5e-11eb-8dab-a7bf3586ff61.png)

Personally I like the second one, since it's so easy to increase the size of a pattern using the mouse wheel (which will increase the nominator by one) and since the "average" user will only experience integer numbers when interacting using the mouse and the "advanced" one can easily have floating point precision by using keyboard.

What do you think?

oddtime commented:
Hi, thanks for asking. I personally like the first option because you can enter the size way more quickly and it occupies less space.

oddtime commented:
Can I ask if we can unblock the size widget when H2 is playing? I had tried this and it seemed to work...

theGreatWhiteShark commented:
> The mouse wheel could work with the QLineEdit too (in the same way as option 2, but just for the numerator)

Yeah. This was my initial thought too. The only thing about the `QLineEdit` that's bothering me is would make the pattern size widget the only element within Hydrogen you can not fully interact with using only the mouse. This makes it feel quite alien.

> Can I ask if we can unblock the size widget when H2 is playing? I had tried this and it seemed to work...

Hmmm. I think this only works when Hydrogen is used as a stand-alone application. When the JACK server is used to sync transport between Hydrogen and other programs this will cause glitches. If any instance of the pattern the users changes size of would occur prior to the current transport position, a change in size would alter this position. So, either Hydrogen has to ask all other application to relocate transport in order to keep the internal one consistent or the transport position within Hydrogen has to be changed.

oddtime commented:
> I think this only works when Hydrogen is used as a stand-alone application. When the JACK server is used to sync transport between Hydrogen and other programs this will cause glitches.

Isn't it the same case when you add a pattern in the pattern sequence while h2 is playing?

theGreatWhiteShark commented:
> Isn't it the same case when you add a pattern in the pattern sequence while h2 is playing?

Excellent point. Can't remember whether I have tried this. How about we move the discussion concerning the dynamic resizing into a separate issue to not loose sight of it once this PR is closed?

oddtime commented:
Ok for separate issue

theGreatWhiteShark commented:
I'm finally coming to an end with this one.

@oddtime After the tweaks to the other input widgets were done I tried both a `QLineEdit` and two spin boxes for the pattern size widget. But the former just felt like a passive element no matter which colors I picked. So, I decided to use the spin boxes instead. It turns out they work very well in constraining the user from inserting invalid input themselves. AFAICS there is no need for a popup and warning triangle anymore since the user can not insert invalid combinations.Could you give it a try and see if it works as expected? Also, I set the minimum value of the numerator to 0.1 since smaller values seem to produce glitches. Is this okay or should there be another minimum? (Note that numbers smaller than 1.0 can only be reached via clicking the widget and entering a number)

oddtime commented:
Hi @theGreatWhiteShark I haven't tried the new commits yet but that's my opinion, I wish to be not unpleasant but I think this:

I would like to have a popping dialog to input the time signature like the one in master (via shortcut or via menu) in addition to the spins for the sake of completeness and rapidity. I think this can be a good compromise.

I remember we discussed the fact of let the user input unsupported denominators (being approximated). I personally have used approximated times it in some of my songs, even if they are not perfect there is a use of them.
There is also a new argument for that: when position become float (double) we will support custom denominators completely, and I hope that we are doing this in v1.2 since tuplet is something many users have been asking for years and if I am not wrong float note positions would solve sync issues with jack too.

oddtime commented:
I've just update it and built and another thing that I miss is the possibility to enter 4.5 / 4 for example, which size is not approximated and can be really found in contemporary music.

I thought you had removed the "unsupported" denominators (5, 7...) instead I see they are present, nice; but I can't enter them via keyboard while I can with the spin buttons.

oddtime commented:
I am not able to change bpm with the keyboard,
Also, when you click on it it seems that you can navigate with left/right arrow keys, but something is not working e.g. when you press backspace.

theGreatWhiteShark commented:
> I would like to have a popping dialog to input the time signature like the one in master (via shortcut or via menu) in addition to the spins 

Fortunately with the new widgets there is no need for popups anymore. All input widgets can handle keyboard input directly.

> for the sake of completeness and rapidity.

I don't get why a QLineEdit is so much more faster. For the line edit the steps to enter a new combination is

1. double click the widget, 2. enter the numerator, 3. enter <kbd>/</kbd>, 4. enter the denominator

for the solution with the two spin boxes it is

1. double click the first spin box, 2. enter the numerator, 3. double click the second spin box, 4. enter the denominator

the difference should be only marginal.

> I remember we discussed the fact of letting the user input unsupported denominators (being approximated). I personally have used approximated times in some of my songs, even if they are not perfect there is a use of them.

I have implemented it in such a way that via arrow buttons and mouse clicks the user can only access integer numerators. But when using the keyboard input float numbers between 0.1 and four times the denominator are supported. So, an explicit user action is required to make the note positioning imprecise. I think this is a good tradeoff since the fractional numerators are an advanced feature and the vast majority of the users will most likely never require them.

>  hope that we are doing this in v1.2 since tuplet is something many users have been asking for years

How exactly does the size of the pattern - or in musical terms the measure - affect the tuplets? Shouldn't they only be concerned with the resolution spin box to the right?

> if I am not wrong float note positions would solve sync issues with jack too.

I'm afraid fractional measures will cause extra bugs and oddities in the JACK sync as the measure is broadcasted in the BBT information and other JACK clients are expected to handle a 1.748/196 measure well.

> I've just update it and built and another thing that I miss is the possibility to enter 4.5 / 4 for example, which size is not approximated and can be really found in contemporary music. I think we should not make the numerator discrete constraining a quantum interval bigger than ticks - hopefully these becoming float at one point.

Did you pulled the most recent changes and performed `sudo make install`? The numerator should be float with two digits after point.

> I thought you had removed the "unsupported" denominators (5, 7...) instead I see they are present, nice; but I can't enter them via keyboard while I can with the spin buttons.

Hahaha, actually I have and this seems to be a bug. These "unsupported" measures should be accessible via the floating point numerator. Do you get an orange rectangle drawn around the spin box once clicking it? This indicates that widget has focus and is ready to receive keyboard input events.

> I am not able to change bpm with the keyboard,
Also, when you click on it it seems that you can navigate with left/right arrow keys, but something is not working e.g. when you press backspace.

That's quite odd. Does anything show up in the error log?

oddtime commented:
mmm... we discussed many of these arguments in pull req #999

> I don't get why a QLineEdit is so much more faster. 

Not so much absolutely, but with lineedit but you don't have to move the cursor in two different places, and with a shortcut this become relatively faster. And since the the dialog code is already pushed so there's no extra work for it, I would keep the feature. ;)
Also i see that it is not sufficient to click on the widget to enter the number via keyboard but you have to select all the text inside the spin box, or cancel. If this is expected then the line edit is even a little faster.


> How exactly does the size of the pattern - or in musical terms the measure - affect the tuplets? Shouldn't they only be concerned with the resolution spin box to the right?

A size of pattern of type float allows tuplet base time signatures, i.e. the unsupported denominators.
then if note position become float they can represent tuplet position, but it is not necessarily linked to the pattern size.

> Did you pulled the most recent changes and performed sudo make install? The numerator should be float with two digits after point.

I think i did it this morning. however I repulled now and did `make && sudo make install`. 
Can you enter one of this values: 4.5/4, 4.51/4 4.52/4, 4.53/4, 4.54/4, 4.55/4 ... ?
The first is accepted while the others are refused. Also, I don't understand but the point separator is not working in input. That's something weird that I am not understanding XD... also I write 5,3 and before even pressing enter it changes to 5.29
My keyboard is Italian but I would like to use the point separator as well (for which I had implemented a way in the parser of the pattern size dialog).
Do yu experience the same strange behaviour?

> Hahaha, actually I have and this seems to be a bug.

But this is not a bug at all actually! Let's think it in the perspective of H2 supporting all denominators... with float size.
And even now that such sizes are not supported you can get a so good approximation without noticing the time difference by ear (in standalone).

> These "unsupported" measures should be accessible via the floating point numerator.

I don't agree: even if the result is the same because at the end H2 use just a number for the size, user may prefer a 4/5 (approximated) to the analogous written in base 4 because it has a musical meaning.


> I'm afraid fractional measures will cause extra bugs and oddities in the JACK sync as the measure is broadcasted in the BBT information and other JACK clients are expected to handle a 1.748/196 measure well.

Ok, I remember just you writing that the core resolution could be in frames rather than in ticks and note positions to become float and that could solve sync issues... I think the best way is to try and see.

oddtime commented:
The grey LCDs for  instr gain and pitch are not interactive but you can select the text. That yellow text selection makes me think you can click on them and enter the value (after all this would be fine)

oddtime commented:
Also I would show another digit in the numerator. Why? If I remember that was for 1.125/4 which is exact...

oddtime commented:
> That's quite odd. Does anything show up in the error log?


nothing... this is me tryng to enter 200 bpm and pressing back space at the end
[file.zip](https://github.com/hydrogen-music/hydrogen/files/6742921/file.zip)

theGreatWhiteShark commented:
> nothing... this is me tryng to enter 200 bpm and pressing back space at the end

Should be fixed now. The timer passed update of the PlayerControl tried to overwrite the value every X ms.

> Also I would show another digit in the numerator. Why? If I remember that was for 1.125/4 which is exact...

Alternatively, I could increase the precision all the way up to float precision and use the tooltip to display the exact value.

> The grey LCDs for instr gain and pitch are not interactive but you can select the text. That yellow text selection makes me think you can click on them and enter the value (after all this would be fine)

Fixed.

theGreatWhiteShark commented:
> I think i did it this morning. however I repulled now and did make && sudo make install.
Can you enter one of this values: 4.5/4, 4.51/4 4.52/4, 4.53/4, 4.54/4, 4.55/4 ... ?
The first is accepted while the others are refused. Also, I don't understand but the point separator is not working in input. That's something weird that I am not understanding XD... also I write 5,3 and before even pressing enter it changes to 5.29
My keyboard is Italian but I would like to use the point separator as well (for which I had implemented a way in the parser of the pattern size dialog).
Do yu experience the same strange behaviour?

That seems to be caused by a similar bug like the one with the BPM spin box. Should be fixed now. Could you give it another try?

theGreatWhiteShark commented:
> Not so much absolutely, but with lineedit you don't have to move the cursor in two different places, and with a shortcut this become relatively faster. And since the dialog code is already pushed so there's no extra work for it, I would keep the feature. ;)
Also I see that it is not sufficient to click on the widget to enter the number via keyboard but you have to select all the text inside the spin box, or cancel. If this is expected then the line edit dialog is even a little faster.

Okay. Let me put it like this. The line edit inside the popup with all its parsing logic was a huge improvement. The idea of this PR is not so much in fine tuning the existing widgets but to make them behave the same way as found in the majority of other applications. Personally, I would prefer to only click once at a spin box in order to change its value. But by diverging from the current default behavior of GUIs one is likely to make them a nuisance or create hidden features instead of improving the UX. Like the resetting of the rotaries using Ctrl+click. It's handy, but at least I used Hydrogen for years before I stumbled across this. A scroll function on a line edit might be a similar one. By deriving from the Qt widgets and adding as few custom stuff as possible Hydrogen is also set up to mature in dignity for the next 20 years since we get all updates of rendering quality, support of new ways of user interaction etc. from newer Qt versions for free.

But I have no hard feelings towards the spin boxes and no general problem with implementing a line edit instead. So, I would suggest we leave the decision to @cme and @mauser as all cards should be on the table by now. Do you agree?

oddtime commented:
> That seems to be caused by a similar bug like the one with the BPM spin box. Should be fixed now. Could you give it another try?

All keyboard entries (BPM and numerator widgets) are working now. Thanks!
What about the comma and point separator? It is still accepting only the local separator, comma in my case.

> Alternatively, I could increase the precision all the way up to float precision and use the tooltip to display the exact value.

Ok with that! I see that now there is enough space to display 4 digits so maybe we could use all them.

oddtime commented:
> Okay. Let me put it like this. The line edit inside the popup with all its parsing logic was a huge improvement. The idea of this PR is not so much in fine tuning the existing widgets but to make them behave the same way as found in the majority of other applications. Personally, I would prefer to only click once at a spin box in order to change its value. But by diverging from the current default behavior of GUIs one is likely to make them a nuisance or create hidden features instead of improving the UX. Like the resetting of the rotaries using Ctrl+click. It's handy, but at least I used Hydrogen for years before I stumbled across this. A scroll function on a line edit might be a similar one. By deriving from the Qt widgets and adding as few custom stuff as possible Hydrogen is also set up to mature in dignity for the next 20 years since we get all updates of rendering quality, support of new ways of user interaction etc. from newer Qt versions for free.
> 
> But I have no hard feelings towards the spin boxes and no general problem with implementing a line edit instead. So, I would suggest we leave the decision to @cme and @mauser as all cards should be on the table by now. Do you agree?

I think that both spins and line edit have their pros and cons. Also, one is not always faster than the other because the spin wins when you just have to increase or decrease the numerator by few units of course, while on the other hand if you have to change both num and den, a single line edit usually wins.
I think that a shortcut like ctrl+T when documented, will not be so hidden as the rotary reset has been (I had the same experience for that, maybe because expected to reset the rotary with double click like in other programs). Such shortcut is generally present in notation softwares like sibelius, whose UX has been imitated by many others like musescore. I knew that Dorico, which was claimed to be the best notation software for modern time signatures, uses to parse a line entry (and accepts any denominator as well).

The thing that I care really the most is the limitation in digits and denominators, for the reasons written in this discussion and in #999.

oddtime commented:
Another question: what if the rotary widgets were deselected (I mean no more yellow underlined and not editable via keyboard) when you click somewhere in the panel?

cme commented:
Hi! I'm catching up, I've just built from the branch. First impressions: yum!

Have you been able to try on hi-dpi display? I'm not sure how it is on a regular display, but there's a few odd proportions, and it looks like (I haven't looked at the implementation yet) the icon SVG's are rendered at  low res currently. Which I'm sure is just a small thing but looks jarring on hidpi. Apart from that it looks great!

<img width="1008" alt="image" src="https://user-images.githubusercontent.com/768466/124171988-2e852680-daa1-11eb-9403-e32614bea9d1.png">

On the size widget issue, I have a suggestion: if I'm starting something from scratch, I'd probably want to type something like "12/18" as "1" "2" "/" "8" -- could "/" in the numerator spin-box move to the denominator?

oddtime commented:
> On the size widget issue, I have a suggestion: if I'm starting something from scratch, I'd probably want to type something like "12/18" as "1" "2" "/" "8" -- could "/" in the numerator spin-box move to the denominator?

I have thought this too, I didn't want to say because I was afraid this is in the category of "hidden" features like ctrl+click.

> Have you been able to try on hi-dpi display?...

My asus laptop has a 4K screen, but since lubuntu 20 doesn't support 4k completely I use mostly HD res. When I switched 4K to see the new GUI in this commits H2 was simply smaller on the screen. Do you know a way to scale it...? Or the gui is still dependent on fixel pixel dimensions? After all many things in the gui (drum pattern editor...) depend on pixel positions which is fixed etc

cme commented:
> My asus laptop has a 4K screen, but since lubuntu 20 doesn't support 4k completely I use mostly HD res. When I switched 4K to see the new GUI in this commits H2 was simply smaller on the screen. Do you know a way to scale it...? Or the gui is still dependent on fixed pixel dimensions? After all many things in the gui depend on pixel positions which is fixed etc for example the margin of drum pattern editor...

It should scale automatically based on the DPI value your X server, but if that's not set, you can force it with

```
env QT_SCALE_FACTOR=2 ./hydrogen
```

Even if the rest of your desktop is out of whack, it's worth it for the hydrogen visuals now :)

oddtime commented:
> It should scale automatically based on the DPI value your X server, but if that's not set, you can force it with

Thanks, it was my fault since I was changing the resolution after H2 had started XD
It looks beautiful at high resolution... ;)
I confirm some widgets are rendered at low res like in your screenshot

oddtime commented:
screen res: 1920x1080
`hydrogen
`![1080](https://user-images.githubusercontent.com/54219881/124193920-f47b4b00-dac7-11eb-912e-a24e6d0c73d0.jpg)


screen res: 3840x2160
`env QT_SCALE_FACTOR=1 hydrogen
`![scale1](https://user-images.githubusercontent.com/54219881/124194467-eda10800-dac8-11eb-9531-a67007014673.jpg)

Is it normal that the proportions are different?

theGreatWhiteShark commented:
> Have you been able to try on hi-dpi display? 

I've recently switched devices and I'm upgrade to luxurious 1920x1080 :D . I used the command-line scaling option of Qt to fine tune the rotary and fader widget. But my tiling window manager allows me to only access the upper left corner of Hydrogen, so I might need additional input to fix these oddities.

> I'm not sure how it is on a regular display, but there's a few odd proportions, and it looks like (I haven't looked at the implementation yet) the icon SVG's are rendered at low res currently. Which I'm sure is just a small thing but looks jarring on hidpi. Apart from that it looks great!

I've noticed that as well but put it aside (and forgot about).  Seems like Qt is first converting the SVG to pixel graphic and afterwards scales this graphic to higher resolution. I'll look into it but I doubt there is more than a global config needed to fix this.

> On the size widget issue, I have a suggestion: if I'm starting something from scratch, I'd probably want to type something like "12/18" as "1" "2" "/" "8" -- could "/" in the numerator spin-box move to the denominator?

Sounds good. The "/" is a label. So, it can have a tooltip hinting the shortcut. (Curiously one can use <kdb>Shift</kbd> + <kdb>Tab</kdb> to switch from the denominator to the numerator. The focus policy seems to be only adopted by plain <kdb>Tab</kdb>. Not quite sure if this is a bug or a feature.)

theGreatWhiteShark commented:
> Is it normal that the proportions are different?

Well. This PR only touches the basic widgets. Most of the layout is still fixed in sized with a couple of dynamic exceptions.


Based on your screenshot I can see some problems
- [ ] the font size of the `LCDDisplay` and `LCDSpinBox` class is to small. This probably should be implemented relative to the widget height to allow for scalability
- [ ] the two checkboxes in the `InstrumentEditor` arn't scaled at all. That's odd. Maybe because we use the plain Qt widget and don't call it's resize() method? I'll check.
- [ ] the font in the beatcounter is too small. That thing is a pain anyway. I'll see if there is a easy fix but I think it's still usable with smaller font.

What about the PreferencesDialog option to change the font size? Is it working properly in the scaled version too?

theGreatWhiteShark commented:
> What about the comma and point separator? It is still accepting only the local separator, comma in my case.

I actually think it's a big improvement that the widget is sensitive towards the locale set by the user. After all, the spin box now works exactly the same as in e.g. Muse or MuseScore. Problem with changing this behavior is that we need to overwrite it entirely and would loose all the nice Qt features.

> Ok with that! I see that now there is enough space to display 4 digits so maybe we could use all them.

The number of digits displayed is due to the Qt internals and judging by the API there seems to be no way to alter this. But I'll double check. Strangely, changing the font size doesn't work properly with this widget.

> Another question: what if the rotary widgets were deselected (I mean no more yellow underlined and not editable via keyboard) when you click somewhere in the panel?

Hmm. Possible but would require some work. Do you expect such a behavior? For context: in MuseScore this appears to be the case but not quite. Clicking the toolbars won't remove focus from the e.g. spin boxes but clicking on dragable widgets, like the palettes, does.

theGreatWhiteShark commented:
> Alternatively, I could increase the precision all the way up to float precision and use the tooltip to display the exact value.

Seems this is not supported by Qt. One can raise the precision and allow for more numbers being inputted but the result is always rounded to the number of digits being displayed. So, the spin box needs to be enlarged to allow for more numbers.

oddtime commented:
> Seems this is not supported by Qt. One can raise the precision and allow for more numbers being inputted but the result is always rounded to the number of digits being displayed. So, the spin box needs to be enlarged to allow for more numbers.

Ok. I haven't seen the code. can it display one more digit without enlarging the box?

oddtime commented:
> I actually think it's a big improvement that the widget is sensitive towards the locale set by the user. After all, the spin box now works exactly the same as in e.g. Muse or MuseScore. Problem with changing this behavior is that we need to overwrite it entirely and would loose all the nice Qt features

damn... that means that keypad won't work without changing the locale

oddtime commented:
> Hmm. Possible but would require some work. Do you expect such a behavior? For context: in MuseScore this appears to be the case but not quite. Clicking the toolbars won't remove focus from the e.g. spin boxes but clicking on dragable widgets, like the palettes, does.

exactly. clicking on the background

oddtime commented:
> Curiously one can use Shift + Tab to switch from the denominator to the numerator. The focus policy seems to be only adopted by plain Tab. Not quite sure if this is a bug or a feature

Is it something already working?

oddtime commented:
editing font size in preferences:
large
![image](https://user-images.githubusercontent.com/54219881/124305520-7ff6e980-db65-11eb-9fca-a07624f7923d.png)

small
![image](https://user-images.githubusercontent.com/54219881/124305582-94d37d00-db65-11eb-95f0-9bc7365ec62e.png)

cme commented:
HiDPI pixmaps -- gorgeous! <3

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/768466/124305985-bcbee280-db5d-11eb-8787-45be2b8205d3.png">

cme commented:
> > Have you been able to try on hi-dpi display? 
> 
> I've recently switched devices and I'm upgrade to luxurious 1920x1080 :D . I used the command-line scaling option of Qt to fine tune the rotary and fader widget. But my tiling window manager allows me to only access the upper left corner of Hydrogen, so I might need additional input to fix these oddities.

Top tip: you can use a vncserver set with a larger resolution to run hydrogen in at higher scale, then connect to that with a vncviewer and it'll let you scroll around arbitrarily, or scale down the vnc desktop to see all.

theGreatWhiteShark commented:
>>   Alternatively, I could increase the precision all the way up to float precision and use the tooltip to display the exact value.
>
> Seems this is not supported by Qt. One can raise the precision and allow for more numbers being inputted but the result is always rounded to the number of digits being displayed. So, the spin box needs to be enlarged to allow for more numbers.

That was actually my bad and Qt is not to blame. I implemented a precision in the numerator up to 16 digits. Per default only two zeros after filled after point (visually more appealing)  but as soon as more digits are entered, all off them will be shown.

> damn... that means that keypad won't work without changing the locale

but isn't that the same with all other applications as well? (You can use `xmodmap` to customize your keyboard layout and replace the comma in the keypad with a dot or the other way around.)


> >  Curiously one can use Shift + Tab to switch from the denominator to the numerator. The focus policy seems to be only adopted by plain Tab. Not quite sure if this is a bug or a feature
>
> Is it something already working?

This a Qt thing and not particular to my code changes (at least not to my knowledge). But I already implemented switching focus back and forth by pressing the <kbd>/</kbd> key.


> > Hmm. Possible but would require some work. Do you expect such a behavior? For context: in MuseScore this appears to be the case but not quite. Clicking the toolbars won't remove focus from the e.g. spin boxes but clicking on dragable widgets, like the palettes, does.
>
> exactly. clicking on the background

I'm not quite sure you got me right here. Removing focus by clicking on a "passive" part of the application (one that can not have focus on its own) is not the default behavior in Qt and neither supported by MuseScore, Muse4, ... . Do you have a feeling for where the default behavior might become an obstacle?

oddtime commented:
Thanks for the commit to jump from num to den with "/"!
There is still an issue as you may have noticed: after pressing "/" I still have to select the den text with the mouse, otherwise the current digit (or digits, I have not checked) remains in the text.
I also notice that backspace doesn't cancel the first digit

Here's an example trying to enter "4/4":
[file.zip](https://github.com/hydrogen-music/hydrogen/files/6758636/file.zip)

theGreatWhiteShark commented:
> Top tip: you can use a vncserver set with a larger resolution to run hydrogen in at higher scale, then connect to that with a vncviewer and it'll let you scroll around arbitrarily, or scale down the vnc desktop to see all.

Of course. Thanks for the hint. Works like a charm.

I have to admit that the state of the GUI is more dire than I expected. There are loads of unnecessary function calls, some things are called over and over again - e.g. switching the pattern via mouse calls all involved functions twice since the update of the pattern length uses the `EVENT_SELECTED_PATTERN_CHANGED` instead of the `EVENT_PATTERN_CHANGED` or `EVENT_PATTERN_MODIFIED` event - and some data flow is quite confusing. But worst of all there are some timers that try to update mostly static stuff over and over again. There could be a gain in performance waiting when porting them to event-based handling. But I haven't touched any of them since this PR is at least 14 times the acceptable size of a PR already :laughing:

oddtime commented:
> I actually think it's a big improvement that the widget is sensitive towards the locale set by the user.

Another comment: the BPM and numerator display use '.' separator (but ',' for the input)
while instrument gain and pitch use ',' separator (but '.' for the input)

oddtime commented:
> I'm not quite sure you got me right here. Removing focus by clicking on a "passive" part of the application (one that can not have focus on its own) is not the default behavior in Qt and neither supported by MuseScore, Muse4,

here's what I mean (musescore):
[file.zip](https://github.com/hydrogen-music/hydrogen/files/6758682/file.zip)

>  ... . Do you have a feeling for where the default behavior might become an obstacle?

After you have changed a knob, it remains selected and it could be changed unintentionally with arrow keys...
But if this means a lot of work now don't worry.

oddtime commented:
> but isn't that the same with all other applications as well? (You can use xmodmap to customize your keyboard layout and replace the comma in the keypad with a dot or the other way around.)

not all apps actually... I have tried now and some accept both (inkscape), some only locale (gimp), some only point (audacity).
thanks for the hint

cme reviewed (commented):
Review request...

...400 files changed. 🤣 

Seriously though -- awesome

theGreatWhiteShark commented:
> Another comment: the BPM and numerator displays use '.' separator (but ',' for the input)
while instrument gain and pitch use ',' separator (but '.' for the input)

That's odd. I'll check.

> After you have changed a knob, it remains selected and it could be changed unintentionally with arrow keys...
But if this means a lot of work now don't worry.

Another problem is that <kbd>space</kbd> is currently not working for toggling playback when one of the widgets is focused. That doesn't feel right.

theGreatWhiteShark commented:
> Review request...
>
> ...400 files changed. rofl

I did hesitate quite a lot before hitting this button. Especially since removing all the pixmaps and adding the SVGs brings the Github page with e.g. it's "jump to file" feature to its knees :laughing: But on the other hand, most changes are just cosmetic and if some button isn't properly wired anymore we will most probably find out before the next release anyway.

theGreatWhiteShark commented on src/gui/src/CommonStrings.h:
```diff
@@ -0,0 +1,212 @@
+	in order to allow for the reusage. This will (some day) make
+	translation work faster.
+
+	Initially I wanted the whole thing as well as the getter methods
+	to be static. But Qt requires the class to be initialized for the
+	translation engine to take effect.*/
```
> (eg. by adding convenience methods in Object) 

I also thought about introducing a `GUIObject` which would be just a wrapper around `H2Core::Object` in `src/gui/src` in order to avoid mixing core and gui stuff. It could have a `getCommonStrings()` protected method (and potentially some debugging function showing basic QWidget properties, like focus, size, tooltip etc.)

theGreatWhiteShark commented on src/gui/src/MainForm.cpp:
```diff
@@ -104,11 +106,34 @@ MainForm::MainForm( QApplication * pQApplication )
 	setFont( font );
 	m_pQApp->setFont( font );
 
+	m_lastWidgetColor = pPref->getDefaultUIStyle()->m_widgetColor;
+	m_lastWidgetTextColor = pPref->getDefaultUIStyle()->m_widgetTextColor;
+	m_lastAccentColor = pPref->getDefaultUIStyle()->m_accentColor;
+	m_lastAccentTextColor = pPref->getDefaultUIStyle()->m_accentTextColor;
+	m_lastToolTipTextColor = pPref->getDefaultUIStyle()->m_toolTipTextColor;
+	m_lastToolTipBaseColor = pPref->getDefaultUIStyle()->m_toolTipBaseColor;
+	m_lastWindowColor = pPref->getDefaultUIStyle()->m_windowColor; 
+	m_lastWindowTextColor = pPref->getDefaultUIStyle()->m_windowTextColor;
+	m_lastSpinBoxSelectionTextColor = pPref->getDefaultUIStyle()->m_spinBoxSelectionTextColor;
+	m_lastSpinBoxSelectionColor = pPref->getDefaultUIStyle()->m_spinBoxSelectionColor;
+	m_lastBaseColor = pPref->getDefaultUIStyle()->m_baseColor;
+	m_lastAlternateBaseColor = pPref->getDefaultUIStyle()->m_alternateBaseColor;
+	m_lastTextColor = pPref->getDefaultUIStyle()->m_textColor;
+	m_lastButtonColor = pPref->getDefaultUIStyle()->m_buttonColor;
+	m_lastButtonTextColor = pPref->getDefaultUIStyle()->m_buttonTextColor;
+	m_lastLightColor = pPref->getDefaultUIStyle()->m_lightColor;
+	m_lastMidLightColor = pPref->getDefaultUIStyle()->m_midLightColor;
+	m_lastDarkColor = pPref->getDefaultUIStyle()->m_darkColor;
+	m_lastMidColor = pPref->getDefaultUIStyle()->m_midColor;
+	m_lastShadowTextColor = pPref->getDefaultUIStyle()->m_shadowTextColor;
+	m_lastHighlightColor = pPref->getDefaultUIStyle()->m_highlightColor;
+	m_lastHighlightedTextColor = pPref->getDefaultUIStyle()->m_highlightedTextColor;
+
```
I just reworked the whole `onPreferencesChanged` thing. Unfortunately, it grew quite naturally and some early assumptions, like "I will do this but I won't go as far as touching the UIStyle", are outdated by now. Should be much cleaner now.

theGreatWhiteShark commented:
> Another problem is that space is currently not working for toggling playback when one of the widgets is focused. That doesn't feel right.

When a `Rotary` or a `Fader` is focused, all common keyboard shortcuts of the GUI - including space for toggling playback - are supported now.


> > Another comment: the BPM and numerator displays use '.' separator (but ',' for the input)
    while instrument gain and pitch use ',' separator (but '.' for the input)
>
> That's odd. I'll check.

Good call. Completely missed this one. Since the input of the `Rotary` and `Fader` is handcrafted and only supports <kdb>.</kdb> it's an inconsistent UX if the input in the spin box differs. Also, all the line edits should display "." as separator too. This should be fixed now.

theGreatWhiteShark commented:
Hey @cme.

For some finishing tweaks of this PR I got the idea to split of everything in `Preferences` that's related to the look of the GUI and store it in an individual class/user-level file. It'll affect everything in the Appearance tab of the PreferencesDialog as of 
5b23182 . This will result in a new .h2theme file type stored in ~/.hydrogen/data/themes. The idea is to allow users to make e.g. a dark theme of Hydrogen (everything required can be found in the PreferencesDialog and no coding at all required). But it's possible I'm too immersed into all the coloring stuff to see I'm going to far on this one. What do you think about the idea?

cme commented:
With the idea that the themes courd be exported / shared etc as well? I think it sounds pretty reasonable, and with a big change in GUI appearance like this I think people would appreciate more ability to tweak than not :)

theGreatWhiteShark commented:
> With the idea that the themes courd be exported / shared etc as well?

Yes. As with the drumkits, the actual state used by Hydrogen is in `hydrogen.conf` and the corresponding settings can be exported to/imported from an `.h2theme` file. This way it would be way more easy to share them e.g. via Github.

theGreatWhiteShark commented:
Hey @cme ,

I'm done with tweaking this PR and would merge it in about a week if you don't object. Afterwards, I would open a dedicated Discussion topic and write to the mailing list to motivate people trying it, help creating a nicer default theme, and to find potential bugs. When the obvious bugs are gone we could call it half way through the current release cycle and bundle master into a 1.2.0-beta. What do you think?

cme commented:
> Hey @cme ,
> 
> I'm done with tweaking this PR and would merge it in about a week if you don't object. Afterwards, I would open a dedicated Discussion topic and write to the mailing list to motivate people trying it, help creating a nicer default theme, and to find potential bugs. When the obvious bugs are gone we could call it half way through the current release cycle and bundle master into a 1.2.0-beta. What do you think?

Oops, sorry I missed this -- yes! Looks great!
